docs/codegen-tour.md

```diff
@@ -0,0 +1,843 @@
+
+In the font file, these values are all represented in [big-endian][endianness]
+byte order. When we access them, we will need to convert them to the native
+endianness of the host platform. We also need to have some set of types which
+exactly match the memory layout (including byte ordering) of the underlying font
+file; this is necessary for us to take advance of zerocopy semantics (see the
+[zerocopy section](#zerocopy) below.)
+
+This leads us to a situation where we require two distinct types for each
+scalar: a native type that we will use in our program logic, and a
+'raw' type that will represent the bytes in the font file (as well as some
+mechanism to convert between them.)
+
+There are various ways we could express this in Rust. The most straightforward
+would be to just have two parallel sets of types: for instance alongside the
+`F2Dot14` type, we might have `RawF2Dot14`, or `F2Dot14Be`. Another option might
+be to have types that are generic over byte-order, such that you end up with
+types like `U16<BE>` and `U16<LE>`.
+
+There's one additional complication around these values: certain types in the
+spec are stored as three bytes, (`uint24` and `Offset24`) but in native code it
+is easier and more efficient to represent these as 32-bit integers. This means
+that the raw and normal types may not have the same size.
+
+I have taken a slightly different approach, which tries to be more ergonomic and
+intuitive to the user, at the cost of having a slightly more complicated
+implementation.
+
+#### `BigEndian<T>` and `Scalar`
+
+Our design has two basic components: a trait, `Scalar` and a type
+`BigEndian<T>`, which look like this:
+
+```rust
+/// A trait for font scalars.
+pub trait Scalar {
+    /// The raw byte representation of this type.
+    type Raw: Copy + AsRef<[u8]>;
+
+    /// Create an instance of this type from raw big-endian bytes
+    fn from_raw(raw: Self::Raw) -> Self;
+    /// Encode this type as raw big-endian bytes
+    fn to_raw(self) -> Self::Raw;
+}
+
+/// A wrapper around raw big-endian bytes for some type.
+#[derive(Clone, Copy, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct BigEndian<T: Scalar>(T::Raw);
+```
+
+The `Scalar` trait handles conversion of a type to and from its raw representation
+(a fixed-size byte array) and the `BigEndian` type is way of representing some
+fixed number of bytes, and associating them with a concrete type; it has `get`
+and `set` methods which read or write the underlying bytes, relying on the
+`from_raw` and `to_raw` methods on Scalar.
+
+This is a compromise. The `Raw` associated type is expected to always be a
+fixed-size byte array; say `[u8; 2]` for a `u16`, or `[u8; 3]` for an `Offset24`.
+
+Ideally, the scalar trait would look like,
+
+```rust
+trait Scalar {
+    const RAW_SIZE: usize;
+    fn from_raw(bytes: [u8; Self::RAW_SIZE]) -> Self;
+    fn to_raw(self) -> [u8; Self::RAW_SIZE];
+}
+```
+
+But this is not currently something we can express with Rust's generics.
```

Comment:
Will it be in future? Is there an RFC or in flight proposal that would permit this?

Reply:
Yes, I believe this will be possible when we get the [`generic_const_exprs`](https://github.com/rust-lang/rust/issues/76560) feature, although that currently has no clear timeline for stabilization.

Reply:
I think that would be worth noting?